oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/backend/MapReducerIgniteAffinityCall.java

```diff
@@ -136,11 +136,23 @@ public boolean isActive() {
     } catch (IgniteFutureTimeoutException e) {
       throw new OSHDBTimeoutException();
     } catch (IgniteException e) {
-      if (e.getCause().getCause() instanceof OSHDBTimeoutException) {
-        throw (OSHDBTimeoutException) e.getCause().getCause();
-      } else {
-        throw e;
-      }
+      // When a timeout happens remotely, the exception might be burried in (few) layers of
+      // "ignite exceptions". This recursively unwinds these and throws the original exception.
+      throw searchRemoteTimeout(e);
+    }
+  }
+
+  /** This recursively unwinds nested ignite exceptions: if an OSHDBTimeoutException is found it is
+   * returned, otherwise the causing exception is thrown.*/
+  private static OSHDBTimeoutException searchRemoteTimeout(IgniteException exception) {
```

Comment:
basicly what searchRemoteTimeout does is unwrapping the IgniteException(s) chain.
Why not just return the unwrapped cause of the IgniteException(s) 
like:
```java
Throwable unwrap(IgniteException e) {
  Throwable cause = e.getCause();
  if (cause instanceof IgniteException && e != cause) {
    return unwrap((IgniteException) cause);
  }
  return cause;
}
```
On the receiver side we can check for a specific cause if needed. In this pull request it is not.

we can even pull this function into an ignite-utility class.

Reply:
`e.getCause()` could be null. [api-doc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Throwable.html#getCause())

so the receiver side should also check for null before re-throwing the cause.

Reply:
> Why not just return the unwrapped cause of the IgniteException(s)

The current version is in my eyes preferable, becuase it is more easily comprehendable (because it is less generic), and we don't need the more flexible "utility" method. If we need one at some point in the future, I would like to keep a `searchRemoteTimeout` method which then would just call the new method and do the null-check, e.g.

```java
private static OSHDBTimeoutException searchRemoteTimeout(IgniteException exception) {
  Throwable unwrapped = ….unwrap(exception);
  if (unwrapped instanceof OSHDBTimeoutException) {
    return (OSHDBTimeoutException) unwrapped;
  } else {
    throw exception;
  }
}
```

Reply:
I like the new version of searchRemoteTimeout using unwrap.
It is much easier to read what it does. :+1: 

But at the moment, as far as I saw it in this pull request, the OSHDBTimeoutException returned from searchRemoteTimeout is throw immediately! 
Do we have a use for the returned OSHDBTimeoutException somewhere else?

Reply:
> Do we have a use for the returned OSHDBTimeoutException somewhere else?

Don't think so. Do you think it would be nicer to write it like this then:

```java
private static OSHDBTimeoutException searchRemoteTimeout(IgniteException exception) {
  Throwable unwrapped = ….unwrap(exception);
  if (unwrapped instanceof OSHDBTimeoutException) {
    throw (OSHDBTimeoutException) unwrapped;
  } else {
    throw exception;
  }
}
```

Reply:
I've changed this now as in the last suggestion, but inlined the `searchRemoteTimeout` method.